src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala

```diff
@@ -69,7 +74,203 @@ trait SubmissionComponent {
+      }))
+    }
+
+    /* saves a submission and associated workflows to a workspace */
+    def save(workspaceContext: SlickWorkspaceContext, submission: Submission): ReadWriteAction[Submission] = {
+
+      def saveSubmissionWorkflows(workflows: Seq[Workflow]) = {
+        DBIO.seq(workflows.map { case (wf) =>
+          workflowQuery.save(workspaceContext, submission.submissionId, wf)
+        }.toSeq: _*)
+      }
+
+      uniqueResult[SubmissionRecord](findSubmissionById(UUID.fromString(submission.submissionId))) flatMap {
+        case None =>
+          val configId = uniqueResult(methodConfigurationQuery.findConfigByName(
+            workspaceContext.workspaceId, submission.methodConfigurationNamespace, submission.methodConfigurationName).map(_.id))
+
+          loadSubmissionEntityId(workspaceContext.workspaceId, submission.submissionEntity) flatMap { eId =>
+            configId flatMap { cId =>
+              (submissionQuery += marshalSubmission(workspaceContext.workspaceId, submission, eId, cId.get))
+            } andThen
+            saveSubmissionWorkflows(submission.workflows)
+          }
+        case Some(submissionRec) =>
+          update(workspaceContext, submission)
+      }
+    } map { _ => submission }
+
+    /* updates a submission (replaces old with new) */
+    def update(workspaceContext: SlickWorkspaceContext, submission: Submission) = {
+      uniqueResult[SubmissionRecord](findSubmissionById(UUID.fromString(submission.submissionId))) flatMap {
+        case None => throw new RawlsException("submission does not exist to be updated: " + submission)
+        case Some(submissionRec) =>
+          delete(workspaceContext, submission.submissionId) andThen
+          save(workspaceContext, submission)
+      }
+    }
+
+    /* deletes a submission and all associated records */
+    def delete(workspaceContext: SlickWorkspaceContext, submissionId: String): ReadWriteAction[Boolean] = {
+
+      def deleteSubmissionWorkflows(submissionId: UUID) = {
+        workflowQuery.findWorkflowsBySubmissionId(submissionId).result.flatMap { recs => DBIO.sequence(recs.map { rec =>
+          workflowQuery.delete(workspaceContext, submissionId.toString, rec.id.toString)
+        })}
+      }
+
+      uniqueResult[SubmissionRecord](findSubmissionById(UUID.fromString(submissionId))) flatMap {
+        case None =>
+          DBIO.successful(false)
+        case Some(submissionRec) => {
+          deleteSubmissionWorkflows(UUID.fromString(submissionId)) andThen
+          findSubmissionById(UUID.fromString(submissionId)).delete
+        } map { count =>
+          count > 0
+        }
+      }
+    }
+
+    /* admin action: lists all active submissions in the system */
+    def listAllActiveSubmissions(): ReadAction[Seq[ActiveSubmission]] = {
+      findActiveSubmissions.result.flatMap(recs => DBIO.sequence(recs.map { rec =>
+        methodConfigurationQuery.loadMethodConfigurationById(rec.methodConfigurationId) flatMap { config =>
```

Comment:
This would translate really well to a for-comprehension.  flatMap + flatMap + flatMap ... + map is ideal for that.